### PR TITLE
WIP: Add DSSE Envelope

### LINKED ISSUE #370: Add support for DSSE
**Description of issue or feature request**:

[Dead Simple Signing Envelope (DSSE)](https://github.com/secure-systems-lab/dsse) is the new signature wrapper recommended for use in in-toto (https://github.com/in-toto/ITE/blob/master/ITE/5/README.adoc) and TUF (pending https://github.com/theupdateframework/taps/pull/138). Naturally, it makes sense for its implementation to live in securesystemslib.

With ITE-5 up for acceptance (https://github.com/in-toto/ITE/issues/22), we are trying to define a transition period for in-toto during which the reference implementation supports generating and verifying metadata in both formats. This time period is dependent at least in part on the DSSE functionality being present in securesystemslib for in-toto to use. We can tentatively define a relative transition period, perhaps a year after in-toto starts supporting both wrappers.

**Current behavior**:

securesystemslib provides some crypto-specific primitives used by TUF and in-toto. Both projects maintain their own implementation of the current signature wrapper.

**Expected behavior**:

securesystemslib continues providing the functionality it currently does and adds DSSE support, allowing the python implementations of both TUF and in-toto to transition to the new wrapper.

cc @SantiagoTorres @joshuagl @mnm678 @trishankatdatadog

### PR DESCRIPTION
**Pull request by PradyumnaKrishna**
Fixes: #370

### Description of the changes being introduced by the pull request:
Pull Request for DSSE Envelope.
- Created a container class named `Envelope` having attributes `payload`, `payloadType`, and `signatures`.
- Built several methods including functions to load and dump the class into JSON/dict.
- Add a method to generate Pre-Auth-Encoding for the envelope.
- Add tests cases to test the `metadata.py` or `Envelope`

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by lukpueh** [comment]
This looks great! Is there anything you want to add, before I take a look? Otherwise, I suggest to *un-draft* the PR.

Also, in our conversation earlier today, you mentioned an alternative design that uses data classes. Do you want comments on that too?

**Comment by PradyumnaKrishna** [comment]
> Also, in our conversation earlier today, you mentioned an alternative design that uses data classes. Do you want comments on that too?

Actually I think `dataclasses` won't have a major affect. Only problem is `asdict` won't convert byte sequence into base64. 
I will push it in another branch.

File: securesystemslib/metadata.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        return (
+            self.payload == other.payload
+            and self.payloadType == other.payloadType
+            and self.signatures == other.signatures
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "Envelope":
+        """Creates a Signature object from its JSON/dict representation.
+
```
Comment (lukpueh) [inline securesystemslib/metadata.py]: Please avoid trailing whitespace
```suggestion

```

File: securesystemslib/metadata.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    Attributes:
+        payload: Arbitrary byte sequence of Serialized Body
+        payloadType: string that identifies how to interpret payload
+        signatures: List of Signature and GPG Signature
+
+    """
+
+    payload: bytes
+    payloadType: str
```
Comment (lukpueh) [inline securesystemslib/metadata.py]: We generally follow [PEP 8 and thus use snake case for variable names](https://peps.python.org/pep-0008/#function-and-variable-names). Can you please change this here and elsewhere in this PR, even though the spec uses camelCase for this field?

```suggestion
    payload_type: str
```

File: securesystemslib/metadata.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+from securesystemslib import formats
+from securesystemslib.signer import Signature
+from securesystemslib.util import b64dec, b64enc
+
+
+class Envelope:
+    """DSSE Envelope.
+
+    Attributes:
+        payload: Arbitrary byte sequence of Serialized Body
```
Comment (lukpueh) [inline securesystemslib/metadata.py]:
```suggestion
        payload: Arbitrary byte sequence of serialized body
```

File: securesystemslib/metadata.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            and self.signatures == other.signatures
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "Envelope":
+        """Creates a Signature object from its JSON/dict representation.
+        
+        Arguments:
+            data:
+                A dict containing a valid payload, payloadType and signatures
```
Comment (lukpueh) [inline securesystemslib/metadata.py]:
```suggestion
            data: A dict containing a valid payload, payloadType and signatures
```

File: securesystemslib/util.py [hunk 5]
```diff
@@ -455,3 +457,51 @@ def digests_are_equal(digest1: str, digest2: str) -> bool:
...
+
+  <Exceptions>
+    TypeError: If "data" is not byte sequence
+
+  <Side Effects>
+    None.
+
+  <Return>
+    base64 string
+  """
```
Comment (lukpueh) [inline securesystemslib/util.py:473]: I appreciate that you adopt the existing code style here (wrt indentation), but I'd say it's okay to deviate from the rest of the file when it comes to docs. Could you please use the same docstring style as above in metadata.py?

File: securesystemslib/util.py [hunk 6]
```diff
@@ -455,3 +457,51 @@ def digests_are_equal(digest1: str, digest2: str) -> bool:
...
+
+  <Return>
+    A byte sequence
+  """
+
+  data = string.encode('utf-8')
+  try:
+    return base64.b64decode(data, validate=True)
+  except binascii.Error:
+    return base64.b64decode(data, altchars='-_', validate=True)
```
Comment (lukpueh) [inline securesystemslib/util.py]: Is it common to to first try without altchars and then with `'-_'`? I suppose you copied this from the [spec poc](https://github.com/secure-systems-lab/dsse/blob/469c6e9fa6c4b7252fb71101084561cfc4cd0fa5/implementation/signing_spec.py#L85-L90), do you know why they wrote it that way? If you do, I'd appreciate a oneliner comment about it.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Impeccable work, @PradyumnaKrishna! 🎉  I only left a few very minor style comments inline, but otherwise this looks great to me.

Regarding style, would you mind running *black*, *isort* and *pylint*  locally over any new modules you add until we fix https://github.com/secure-systems-lab/securesystemslib/issues/243?  

see https://github.com/secure-systems-lab/securesystemslib/issues/243#issuecomment-1156374257 for details

Also some test would be great. But we can do that in a follow-up PR.

**Comment by PradyumnaKrishna** [comment]
@lukpueh, I guess the indent size is 2 as almost all files follow it except [signer.py](https://github.com/secure-systems-lab/securesystemslib/blob/0c8f834a1f4542a7b2dfa11caa0bdccbb3e5eb94/securesystemslib/signer.py) which uses 4 space of indent. Do I have to change the indent size in `metadata.py` that I have created.

**Comment by lukpueh** [comment]
> Do I have to change the indent size in metadata.py that I have created.

No, it's okay to use 4 in `metadata.py`, which is new, but 2 in `util.py`, which already uses 2.

File: securesystemslib/util.py (hunk 6 above)
Comment (PradyumnaKrishna) [reply securesystemslib/util.py]: I found that urlsafe base64 have - instead of + and _ instead of / that's why its used. so altchars replaces it and try to decode.

**Comment by PradyumnaKrishna** [comment]
By the end of this week I will try to add tests for the implemented methods and then it will be good to merge. Afterwards I will look into sign and verification of `payload` and generating `signatures`.

File: tests/test_metadata.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            "payloadType": "http://example.com/HelloWorld",
+            "signatures": [cls.signature_dict],
+        }
+        cls.pae = b"DSSEv1 29 http://example.com/HelloWorld 11 hello world"
+
+    def test_envelope_from_to_dict(self):
+        """Test envelope to_dict and from_dict methods"""
+
+        # create envelope object from its dict.
+        envelope_obj = Envelope.from_dict(copy.deepcopy(self.envelope_dict))
```
Comment (PradyumnaKrishna) [inline tests/test_metadata.py]: `envelope_dict` contains a nested dict and these tests changes the signature, so I have to use `deepcopy`.

**Review by PradyumnaKrishna — COMMENTED** [review]
I am not sure about GitHub actions can work in forks, please check if they can enable in this repository.

File: securesystemslib/util.py (hunk 5 above)
Comment (lukpueh) [reply securesystemslib/util.py:473]: Ping

File: securesystemslib/util.py (hunk 6 above)
Comment (lukpueh) [reply securesystemslib/util.py]: Thanks for adding a comment

**Comment by lukpueh** [comment]
> I am not sure about GitHub actions can work in forks, please check if they can enable in this repository.

I just enabled them for the fork. Not sure how to trigger the workflow for the existing PR. Maybe pushing a new commit does it?

File: securesystemslib/util.py (hunk 5 above)
Comment (PradyumnaKrishna) [reply securesystemslib/util.py:473]: Actually I was following the docstring used by [`signer.py`](https://github.com/secure-systems-lab/securesystemslib/blob/0c8f834a1f4542a7b2dfa11caa0bdccbb3e5eb94/securesystemslib/signer.py#L1). Do I have to add all the fields like `Purpose`, `Arguments`, `Exceptions`, `Side Effects` and `Returns` in class methods because many functions don't have any arguments or exceptions. or have to use `<>` for specifying field in docs.

**Comment by PradyumnaKrishna** [comment]
> Maybe pushing a new commit does it?

I think reopening this PR would work, let me try.

File: securesystemslib/util.py (hunk 5 above)
Comment (lukpueh) [reply securesystemslib/util.py:473]: Our code style guide has some [notes about docstrings](https://github.com/secure-systems-lab/code-style-guidelines/blob/master/python.md#code-documentation).

TLDR: In new functions/methods/classes, we should use the Google docstring-style, even if the new code is added to an existing module that otherwise uses the old docstring style, which you used here.

Context: The advantage of the Google docstring-style is that there is a Sphinx-parser that can convert it for Readthedocs. And that it is generally more compact, e.g. the `<Purpose>` header is omitted because it is implicit and empty sections are also omitted  (often times this is the `<Side Effects>` section).

Please take a quick look at the style guide linked above and let me know if you have more questions.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Excellent PR and I really appreciate the commit discipline! I'm inclined to just merge it, but maybe you can still change the docstring style for the base64 function (see inline comment)?

Regrading pre-authentication encoding, you will likely also need a deserialization function. You can add it in a follow-up PR. But maybe it makes sense to already spend a thought on its name and whether `pae` will then still make sense for the serialization function implemented here.

File: securesystemslib/metadata.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        for signature in data["signatures"]:
+            if formats.GPG_SIGNATURE_SCHEMA.matches(signature):
+                raise NotImplementedError
+
+            if formats.SIGNATURE_SCHEMA.matches(signature):
+                signatures.append(Signature.from_dict(signature))
+
+            else:
+                raise TypeError(
+                    f"expected type 'Signature', got {type(signature).__name__}"
```
Comment (lukpueh) [inline securesystemslib/metadata.py]: The error message is not accurate, you expected a dictionary that can be parsed using `Signature.from_dict`. An error message like `f"unrecognized signature {signature}"` should be fine.

Also, a more general `ValueError` seems more appropriate here, because the expected Python type i.e. `dict`, might be correct, but its value is not.

**Review by lukpueh — CHANGES_REQUESTED** [review]

File: securesystemslib/metadata.py (hunk 8 above)
Comment (PradyumnaKrishna) [reply securesystemslib/metadata.py]: Already fixed 😄

**Review by lukpueh — APPROVED** [review]
